#  هدف پروژه:
## توضیح کلی پروژه و تعریف صورت مسئله:
### گندالف خاکستری و یاران حلقه

گندالف خاکستری با هدف رهبری و هدایت یاران حلقه مأموریت یافته با ارباب تاریکی سائورون مقابله کند. برای این کار
باید یاران حلقه را از بین اورک‌ها که ارتش سائورون هستند عبور دهد و هر یک را به یک موقعیت مشخص برساند و در انتها
خود به گاندور می‌رود. گندالف نقشه‌ای شطرنجی دارد که شامل موارد زیر است:
+  نقطه شروع گندالف 
+  موقعیت هر یک از یاران حلقه 
+  موقعیت اورک‌ها 
+  محلی که هر یار حلقه باید در آن قرار گیرد 
+  موقعیت نهایی (گاندور)   
 هر یک از اورک‌ها یک درجه نظامی دارد که براساس آن خانه‌هایی با فاصله من‌هتنی n را نظارت می‌کند که گندالف و یاران حلقه نمی‌توانند از آن عبور کنند. چون قرار است گندالف در شب از بین اورک‌ها عبور کند و آن‌ها خواب هستند، او می‌تواند از خانه‌هایی که آن‌ها نظارت می‌کنند عبور کند به این شرط که آن‌ها را بیدار نکند. بعد از عبور n خانه در محوطه تحت نظارت اورك‌ها، آن‌ها بيدار مي‌شوند. بعد از خارج شدن از محدوده تحت نظارت اورک‌ها می‌توان به آن محدوده بدون در نظر گرفتن حرکات قبلی بازگشت. گندالف در هر حرکت می‌تواند به یکی از چهار خانه مجاور خانه فعلی (بالا، پایین، چپ و راست) حرکت کند به شرطی که خانه مقصد خارج از محدوده نقشه نباشد و همچنین اورکی در آن وجود نداشته باشد. گندالف با گذشتن از خانه‌ای که در آن یاران حلقه قرار دارند، او را برمی‌دارد و پس از رسیدن به مکان قرارگیری یار حلقه، او را بدون صرف زمان در آنجا مستقر می‌کند. گندالف تنها می‌تواند در هر زمان تنها یکی از یاران حلقه را به همراه داشته باشد. هیچ‌گاه موقعیت یاران و محلی که باید در آن قرار گیرند در محدوده نظارت اورک‌ها نیست و دو محدوده تحت نظارت اورک‌ها اشتراک ندارند.

### فرمت ورودی
اطلاعات اولیه در یک فایل به فرمت زیر در اختیار برنامه قرار خواهد گرفت:
+ در سطر اول به ترتیب n و m آمده است که n تعداد سطرهای صفحه و m تعداد ستون‌های آن می‌باشد.
+ دو سطر بعدی به ترتیب مختصات نقطه شروع و پایان را نشان می‌دهد (ابتدا شماره سطر از بالا سپس شماره ستون از چپ).
+ در سطر بعدی k و l آمده است که k تعداد اورک‌ها و l تعداد یاران حلقه را نشان می‌دهد.
+ در k سطر بعدی x و y و c آمده است که x شماره‌ی سطر و y شماره‌ی ستون و c درجه نظامی اورک یا فاصله من‌هتنی تحت نظارت یک اورک است.
+ در l سطر بعدی مکان فعلی یاران حلقه به ترتیب آمده است.
+ در l سطر بعدی به همان ترتیب قبلی مکانی که یاران حلقه باید در آن‌ها قرار بگیرند آمده است. توجه داریم هر یار حلقه باید در مکان مشخصی قرار بگیرد. یعنی یار حلقه iام باید در مکان iام قرار بگیرد.

### شرح پیاده‌سازی
کتابخانه‌های استفاده شده در حل مسئله عبارت است از:
+  deque.Collection :برای پیاده‌سازی صف در الگوریتم BFS و IDS
+  time :برای محاسبه زمان
+  heapq :برای پیاده‌سازی صف اولویت در الگوریتم \*A و \* weighted A

In [2]:
import heapq
import math
import time
from collections import deque

### نحوه مدل کردن مسئله
ابتدا یک کلاس Node تعریف کرده‌ایم که شامل ویژگی‌های زیر می‌شود:
+ state :استیت آن node
+ parent :والد آن نود در درخت جست و جو
+ action :حرکتی که باعث بوجود آمدن این node شده است
+ path_cost :هزینه رسیدن از استیت اولیه به این node
+ fn :مقدار تابع (n)f در الگوریتم‌های \*A برای این node

In [3]:
class Node:
    def __init__(self, state, parent, action, path_cost):
        self.state = state
        self.parent = parent
        self.action = action
        self.path_cost = path_cost
        self.fn = 0

    def __lt__(self, other):
        return self.fn < other.fn

    def __gt__(self, other):
        return self.fn > other.fn

دو عملگر lt (کوچکتر از) و gt (بزرگتر از) برای دو تابع heappop و heappush در پیاده‌سازی صف اولویت برای الگوریتم \*A تعریف شده‌اند.
در مدل‌سازی مسئله، 4 حرکت (action) بالا (U)، پائین(D)، چپ(L) و راست(R) در نظر گرفته شده است.
هر استیت بصورت یک tuple سه‌تایی تعریف شده است:
1. شامل مکان فرد در نقشه شطرنجی که بصورت یک tuple به شکل (y,x) درنظر گرفته شده است.
2. مکان اولیه و نهایی یاران حلقه که هنوز در مکان نهایی خود قرار نگرفته‌اند که بصورت tuple of tuple درنظر گرفته شده‌اند.
3. مكان نهايي يار حلقه‌اي كه همراه گندالف است به صورت يك tuple دوتايي مي‌باشد.

استیت اولیه یا state initial شامل مکان اولیه گندالف در بخش اول، مكان اوليه و نهايي همه ياران حلقه در نقشه شطرنجي در بخش دوم و یک tuple خالی در بخش سوم استیت می‌باشد.
استیت هدف یا state goal استیتی است که در آن گندالف به نقطه پایان رسیده است و تمامی ياران حلقه به مكان نهايي خود رسيده اند. بنابراين بخش دوم و سوم tuple استیت هدف شامل tupleهای خالی می‌شود.

### توضیح توابع

تابع actions :در این تابع بررسی می‌شود که گندالف در هر استیت چه حرکاتی می‌تواند انجام دهد و لیستی از این حرکات ممکن بازگردانده می‌شود (درصورتی که گندالف امکان همراهي يكي از ياران حلقه را در محل فعلی داشته باشد این کار نيز در این تابع انجام می‌شود).

In [4]:
def actions(table, state, n, m):
    location = state[0]
    fellowships = state[1]
    fellowship_with_gandalf_end_location = state[2]
    possible_actions = []
    if len(fellowship_with_gandalf_end_location) == 0 :
        for i in range(len(fellowships)):
            if location == fellowships[i][0]:
                fellowship_with_gandalf_end_location = fellowships[i][1]
                fellowships = fellowships[:i] + fellowships[i + 1:]
                break
    if len(fellowship_with_gandalf_end_location) != 0 :
        if location == fellowship_with_gandalf_end_location :
            fellowship_with_gandalf_end_location = ()
    # Move to Right
    if location[1] + 1 < m:
        if table[location[0]][location[1] + 1] != '*':
            possible_actions.append('R')
    # Move to Left
    if location[1] - 1 >= 0:
        if table[location[0]][location[1] - 1] != '*':
            possible_actions.append('L')
    # Move to Up
    if location[0] - 1 >= 0:
        if table[location[0] - 1][location[1]] != '*':
            possible_actions.append('U')
    # Move to Down
    if location[0] + 1 < n:
        if table[location[0] + 1][location[1]] != '*':
            possible_actions.append('D')
    return possible_actions, fellowships, fellowship_with_gandalf_end_location

تابع create_new_node :در این تابع، باتوجه به حرکتی که انجام می‌شود، node جدید (گره فرزند) در درخت جست‌و‌جو تشکیل
داده می‌شود.

In [5]:
def create_new_node(node, action, index):
    parent = node
    path_cost = node.path_cost + 1
    location = node.state[0]
    fellowships = node.state[1]
    fellowship_with_gandalf_end_location = node.state[2]

    if action == 'R':
        location = list(location)
        location[1] = location[1] + 1
    if action == 'L':
        location = list(location)
        location[1] = location[1] - 1
    if action == 'U':
        location = list(location)
        location[0] = location[0] - 1
    if action == 'D':
        location = list(location)
        location[0] = location[0] + 1
    state = tuple(location), fellowships, fellowship_with_gandalf_end_location
    new_node = Node(state, parent, action, path_cost)
    return new_node

تابع goal_test :برای بررسی اینکه الگوریتم سرچ موردنظر به استیت هدف رسیده است یا خیر.

In [6]:
def goal_test(state, end_point):
    location = state[0]
    fellowships = state[1]
    fellowships_of_the_ring_end_location = state[2]
    return location[0] == end_point[0] and location[1] == end_point[1] and \
           len(fellowships_of_the_ring_end_location) == 0 and len(fellowships) == 0

تابع to_find_path :براي پیدا کردن مسیر رسیدن به استیت هدف با استفاده از node آخر و دسترسی به اجداد آن node.

In [7]:
def to_find_path(node):
    temp = []
    path = ''
    while node.parent != -1:
        temp.append(node.action)
        node = node.parent
    temp.reverse()
    return path.join(temp)

تابع get_input :براي گرفتن ورودي از كنسول بكار مي‌رود.

In [8]:
def get_input():
    n, m = [int(t) for t in input().split()]
    start_point = tuple([int(t) for t in input().split()])
    end_point = tuple([int(t) for t in input().split()])
    k, l = [int(t) for t in input().split()]

    orcs = []
    for i in range(k):
        x, y, grade = [int(t) for t in input().split()]
        location = (x, y)
        orcs.append((location,grade))

    fellowships_of_the_ring_start_location = []
    for i in range(l):
        fellowships_of_the_ring_start_location.append([int(t) for t in input().split()])

    fellowships_of_the_ring_locations = []
    for i in range(l):
        end_location = tuple([int(t) for t in input().split()])
        locations = tuple(fellowships_of_the_ring_start_location[i]),end_location
        fellowships_of_the_ring_locations.append(locations)
    return n, m, start_point, end_point, orcs, tuple(fellowships_of_the_ring_locations)

تابع init_table :براي ايجاد صفحه شطرنجي بكار مي‌رود. اين تابع با توجه به درجه‌ي نظامي اورك و اين واقعيت كه گندالف به هيچ عنوان نمي‌تواند حريم n-1 من‌هتني اورك شود صفحه شطرنجي را تشكيل مي‌دهد.

In [9]:
def init_table(n, m, orcs):
    table = [['_']*m for i in range(n)]
    for i in range(len(orcs)):
        x, y = orcs[i][0]
        grade = orcs[i][1]
        for j in range(-grade, grade):
            for k in range(-grade, grade):
                if abs(j) + abs(k) < grade:
                    if (x + j < n and x + j >= 0 and y + k < m and y + k >= 0):
                        table[x + j][y + k] = '*'
    #print_table(table, n, m)
    return table

تابع print_table :براي رسم صفحه شطرنجي بكار مي‌رود كه اين صفحه شطرنجي موقعيت هايي را كه گندالف به هيچ عنوان نمي‌تواند در آن قرار بگيرد بصورت \* نمايش مي‌دهد.

In [10]:
def print_table(table, n ,m) :
    for i in range(n):
        for j in range(m):
            print(table[i][j], end=' ')
        print()

تابع print_output :براي چاپ كردن خروجي بكار مي‌رود.

In [11]:
def print_output(tic, toc, node, number_of_unique_states, number_of_states):
    print(f'execution time:  {(toc - tic)*1000:.6f} ms')
    print("number_of_states: ", number_of_states)
    print("number_of_unique_states: ", number_of_unique_states)
    print("path cost: ", node.path_cost)
    print("path: ", to_find_path(node))

### الگوريتم و تابع BFS

در این الگوریتم nodeها به ترتیب سطح در درخت جست‌و‌جو و بصورت FIFO بسط داده می‌شوند. برای پیاده‌سازی مجموعه frontier
از صف (deque) و برای پیاده‌سازی مجموعه explored از ساختمان داده set استفاده شده است.  
در این الگوریتم هزینه انجام هر حرکت برابر 1 است در نتیجه جواب نهایی بهینه خواهد بود.

In [12]:
def BFS(table, state, n, m , end_point):
    number_of_states = 1
    node = Node(state, -1, '', 0)
    if goal_test(state, end_point) :
        return node, 1, 1
    frontier = deque()
    frontier.append(node)
    explored = set()
    explored.add(node)
    while bool(frontier) :
        node = frontier.popleft()
        possible_actions, fellowships, fellowship_with_gandalf_end_location = actions(table, node.state, n, m)
        node.state = (node.state[0], fellowships, fellowship_with_gandalf_end_location)
        for action in possible_actions:
            child_node = create_new_node(node, action, fellowship_with_gandalf_end_location)
            number_of_states += 1
            if goal_test(child_node.state, end_point):
                return child_node, len(explored), number_of_states
            if not (child_node.state in explored) :
                frontier.append(child_node)
                explored.add(child_node.state)
    return 0, len(explored), number_of_states

### الگوریتم IDS و توابع IDS و DLS

این الگوریتم همانند الگوریتم DFS، در عمق درخت جست‌و‌جو پیش می‌رود و nodeها را بصورت LIFO بسط می‌دهد با این تفاوت که در
این الگوریتم عمق پیشروی در درخت از مقدار صفر شروع می‌شود و الگوریتم DFS تنها تا آن عمق پیش می‌رود(DLS). درصورت نیافتن
جواب در این عمق، عمق یکی اضافه می‌شود و الگوریتم DFS دوباره تا عمق مشخصی اجرا می‌شود. این کار تا زمانی ادامه پیدا می‌کند که تابع goal_test مقدار true برگرداند و الگوریتم سرچ به استیت هدف رسیده باشد.   
در این الگوریتم نیز چون هزینه تمامی حرکت‌ها با یکدیگر برابر و برابر 1 است، جواب نهایی بهینه خواهد بود.

In [13]:
def DLS(table, state, n, m, end_point, limit, unique_states):
    number_of_states = 1
    node = Node(state, -1, '', 0)
    unique_states.add(node.state)
    if goal_test(state, end_point) :
        return node, 1
    frontier = deque()
    frontier.append(node)
    seen_states = set()
    seen_states.add((node.state, 0))
    while bool(frontier) :
        node = frontier.pop()
        possible_actions, fellowships, fellowship_with_gandalf_end_location = actions(table, node.state, n, m)
        node.state = node.state[0], fellowships, fellowship_with_gandalf_end_location
        if goal_test(node.state, end_point):
            return node, number_of_states
        if node.path_cost < limit:
            for action in possible_actions:
                child_node = create_new_node(node, action, fellowship_with_gandalf_end_location)
                number_of_states += 1
                if (node.path_cost < limit):
                    if not ((child_node.state, child_node.path_cost) in seen_states) :
                        frontier.append(child_node)
                        unique_states.add(child_node.state)
                        for i in range(node.path_cost, limit):
                            seen_states.add((child_node.state, i))
    return 0, number_of_states

In [14]:
def IDS(table, state, n, m, end_point):
    depth = 0
    unique_states = set()
    number_of_states = 0
    result = 0
    while result == 0:
        result, i = DLS(table, state, n, m, end_point, depth, unique_states)
        number_of_states += i
        depth += 1
    return result, len(unique_states), number_of_states

### الگوریتم \*A  و  \* weighted A و تابع AStar

در ابتدا به معرفی heuristicهای تعریف شده پرداخته ايم:
+ h1 :تعداد ياران حلقه باقي‌مانده که هنوز در محل درست خود قرار نگرفته‌اند.
+ h2 :فاصله من‌هتن گندالف تا نزدیک‌ترین يار روي صفحه شطرنجي که هنوز به محل درست خود نرفته است.
+ h3 :فاصله اقليدسي تا نزديك‌ترين يار صفحه شطرنجي که هنوز به محل درست خود نرفته است.

هر سه heuristic تعریف شده admissible هستند زیرا به ازای هر node فاصله رسیدن به استيت هدف مقدار بیشتری (یا مساوی)
نسبت به h1 یا h2 يا h3 دارد.
h1 تابعي consistent است زیرا در مسیر رسیدن به استیت هدف مقدار g(n) + h(n) = f(n) به ازای تمامی nodeها مقدار غیرنزولی دارد چرا که مقدار (n)g حداکثر یکی اضافه می‌شود و مقدار (n)h نیز حداکثر یکی کم می‌شود.  
حال به توضیح الگوریتم‌ها پرداخته مي‌شود. در الگوریتم \*A ابتدا nodeها در یک صف اولویت وارد می‌شوند و سپس به ترتیب صعودی مقدار (n)f برای تمامی nodeها، مرتب خواهند شد و پس از آن بصورت FIFO بسط داده می‌شوند. مقدار (n)f در هر node برابر g(n) + h(n) خواهد بود که مقدار (n)g برابر هزینه واقعی از initial state تا آن node و (n)h برابر فاصله تخمینی از این node تا goal state است.  
حال اگر مقدار (n)h به ازای هر نود در اجرا الگوریتم \*A در یک عدد ثابت آلفا ضرب شود، این الگوریتم به الگوریتم \* weighted A تبدیل می‌شود. بدلیل ضرب کردن یک مقدار در heuristic به ازای هر node ، ممکن است admissible بودن heuristic از بین برود و جواب نهایی الگوریتم بهینه نباشد.

In [15]:
def AStar(table, state, n, m, end_point, alpha, h):
    number_of_states = 1
    node = Node(state, -1, "", 0)
    if goal_test(node.state, end_point):
        return node, 1, 1
    frontier = []
    node.fn = node.path_cost + alpha * h(n, m, node)
    frontier.append(node)
    heapq.heapify(frontier)
    explored = set()
    explored.add(node.state)

    while bool(frontier) :
        node = heapq.heappop(frontier)
        possible_actions, fellowships, fellowship_with_gandalf_end_location = actions(table, node.state, n, m)
        node.state = (node.state[0], fellowships, fellowship_with_gandalf_end_location)
        for action in possible_actions:
            child_node = create_new_node(node, action, fellowship_with_gandalf_end_location)
            child_node.fn = child_node.path_cost + alpha * h(n, m, child_node)
            number_of_states += 1
            if goal_test(child_node.state, end_point):
                return child_node, len(explored), number_of_states
            if not (child_node.state in explored) :
                heapq.heappush(frontier, child_node)
                explored.add(child_node.state)
    return 0, len(explored), number_of_states

In [16]:
def h1(n, m, node):
    heuristic = len(node.state[1])
    if len(node.state[2]) != 0 :
        heuristic += 1
    return heuristic

In [17]:
def h2(n, m, node):
    location = state[0]
    fellowships = state[1]
    min = n + m
    for i in range(len(fellowships)):
        temp = abs(location[0] - fellowships[i][0][0]) + abs(location[1] - fellowships[i][0][1])
        if temp < min :
            min = temp
    return min

In [18]:
def h3(n, m, node):
    location = state[0]
    fellowships = state[1]
    min = n + m
    for i in range(len(fellowships)):
        temp = math.sqrt((location[0] - fellowships[i][0][0])**2 + (location[1] - fellowships[i][0][1])**2)
        if temp < min :
            min = temp
    return min

Test case 00 :

In [19]:
if __name__ == '__main__':
    n, m, start_point, end_point, orcs, fellowships_of_the_ring_locations = get_input()
    table = init_table(n, m, orcs)

    state = start_point, fellowships_of_the_ring_locations, ()

10 10
0 0
9 9
6 4
2 4 2
2 8 1
4 0 2
6 3 1
7 1 1
7 7 3
0 7
9 1
4 5
3 2
9 1
0 9
4 9
8 0


In [20]:
print("******************BFS******************")
tic = time.time()
node, number_of_unique_states, number_of_states = BFS(table, state, n, m, end_point)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

******************BFS******************
Path not found!!


In [21]:
print("******************IDS******************")
tic = time.time()
node, number_of_unique_states, number_of_states = IDS(table, state, n, m, end_point)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

******************IDS******************


KeyboardInterrupt: 

In [ ]:
print("****************AStar-h1***************")
tic = time.time()
node, number_of_unique_states, number_of_states = AStar(table, state, n, m, end_point, 1, h1)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

In [51]:
print("****************AStar-h2***************")
tic = time.time()
node, number_of_unique_states, number_of_states = AStar(table, state, n, m, end_point, 1, h2)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

****************AStar-h2***************
execution time:  218.066454 ms
number_of_states:  22762
number_of_unique_states:  7221
path cost:  48
path:  RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDURDURDRUDURDRRRRRR


In [52]:
print("*******weighted AStar-h1-alpha=5*******")
tic = time.time()
node, number_of_unique_states, number_of_states = AStar(table, state, n, m, end_point, 5, h1)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

*******weighted AStar-h1-alpha=5*******
execution time:  4.414558 ms
number_of_states:  795
number_of_unique_states:  293
path cost:  48
path:  RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR


In [53]:
print("*******weighted AStar-h2-alpha=5*******")
tic = time.time()
node, number_of_unique_states, number_of_states = AStar(table, state, n, m, end_point, 5, h2)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

*******weighted AStar-h2-alpha=5*******
execution time:  178.462505 ms
number_of_states:  22762
number_of_unique_states:  7221
path cost:  48
path:  RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDURDURDRUDURDRRRRRR


In [54]:
print("*******weighted AStar-h3-alpha=5*******")
tic = time.time()
node, number_of_unique_states, number_of_states = AStar(table, state, n, m, end_point, 5, h3)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

*******weighted AStar-h3-alpha=5*******
execution time:  262.708426 ms
number_of_states:  22762
number_of_unique_states:  7221
path cost:  48
path:  RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDURDURDRUDURDRRRRRR


In [55]:
print("*****weighted AStar-h1-alpha=10000*****")
tic = time.time()
node, number_of_unique_states, number_of_states = AStar(table, state, n, m, end_point, 10000, h1)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

*****weighted AStar-h1-alpha=10000*****
execution time:  6.298065 ms
number_of_states:  760
number_of_unique_states:  282
path cost:  48
path:  RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR


|            |Path cost|Number of states|Number of unique states|Execution time|
|:-----------|:-------:|:--------------:|:---------------------:|:------------:|
|BFS         |48       |22507           |7125                   |100.15        |
|IDS         |48       |370581          |7263                   |2469.25       |
|A\*         |48       |15565           |5240                   |101.18        |
|weighted A\*|52       |760             |282                    |6.69          |

Test case 01 :

In [56]:
if __name__ == '__main__':
    n, m, start_point, end_point, orcs, fellowships_of_the_ring_locations = get_input()
    table = init_table(n, m, orcs)

    state = start_point, fellowships_of_the_ring_locations, ()


9 9
0 0
8 8
6 3
1 3 1
1 7 2
2 1 1
5 3 1
6 1 1
6 6 2
2 4
3 6
4 8
8 1
8 3
8 5


In [57]:
print("******************BFS******************")
tic = time.time()
node, number_of_unique_states, number_of_states = BFS(table, state, n, m, end_point)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

******************BFS******************
execution time:  21.154404 ms
number_of_states:  3736
number_of_unique_states:  1267
path cost:  52
path:  RRRRDDLLDDDDDLDRRRUUURURULLDDDLDDRRRRRUUUUDDDLDLLRRR


In [58]:
print("******************IDS******************")
tic = time.time()
node, number_of_unique_states, number_of_states = IDS(table, state, n, m, end_point)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

******************IDS******************
execution time:  2943.961143 ms
number_of_states:  483105
number_of_unique_states:  1261
path cost:  52
path:  DRRDRRDDDDDDLLLURUUUURRRRDLDLDDDLUURUURRRRDDDDLLLRRR


In [59]:
print("****************AStar-h1***************")
tic = time.time()
node, number_of_unique_states, number_of_states = AStar(table, state, n, m, end_point, 1, h1)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

****************AStar-h1***************
execution time:  37.859440 ms
number_of_states:  3684
number_of_unique_states:  1257
path cost:  52
path:  RRRRDDLDLDDDDDLURUUURRURRLLDDDDDLRRRRRUUUUDDDLDLLRRR


In [60]:
print("*****weighted AStar-h1-alpha=10000*****")
tic = time.time()
node, number_of_unique_states, number_of_states = AStar(table, state, n, m, end_point, 10000, h1)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

*****weighted AStar-h1-alpha=10000*****
execution time:  14.994383 ms
number_of_states:  1260
number_of_unique_states:  467
path cost:  52
path:  RRRRDDLDLDDDDDLRRUURUURRUDLDLDLDDUURURURRRDDDLDLLRRR


|            |Path cost|Number of states|Number of unique states|Execution time|
|:-----------|:-------:|:--------------:|:---------------------:|:------------:|
|BFS         |52       |3736            |1267                   |29.80         |
|IDS         |52       |483105          |1261                   |3349.18       |
|A\*         |52       |3684            |1257                   |34.08         |
|weighted A\*|52       |1260            |467                    |12.34         |

Test case 02 :

In [61]:
if __name__ == '__main__':
    n, m, start_point, end_point, orcs, fellowships_of_the_ring_locations = get_input()
    table = init_table(n, m, orcs)

    state = start_point, fellowships_of_the_ring_locations, ()

8 8
0 0
7 7
3 2
3 1 2
2 6 1
6 4 2
0 7
7 1
5 0
5 6


In [70]:
print("******************BFS******************")
tic = time.time()
node, number_of_unique_states, number_of_states = BFS(table, state, n, m, end_point)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

******************BFS******************
execution time:  8.559942 ms
number_of_states:  992
number_of_unique_states:  350
path cost:  34
path:  RRRRRRRLLLLDDDDLDLLRDDRUURURRRDRDD


In [75]:
print("******************IDS******************")
tic = time.time()
node, number_of_unique_states, number_of_states = IDS(table, state, n, m, end_point)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

******************IDS******************
execution time:  231.466532 ms
number_of_states:  36067
number_of_unique_states:  359
path cost:  34
path:  RRRRRRRDDDDLLLLDLLLDDRUURURRRDRDDR


In [80]:
print("****************AStar-h1***************")
tic = time.time()
node, number_of_unique_states, number_of_states = AStar(table, state, n, m, end_point, 1, h1)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

****************AStar-h1***************
execution time:  7.537842 ms
number_of_states:  986
number_of_unique_states:  346
path cost:  34
path:  RRRRRRRLDLDDDLLDLLLDDRRUURURRRDDRD


In [87]:
print("*****weighted AStar-h1-alpha=10000*****")
tic = time.time()
node, number_of_unique_states, number_of_states = AStar(table, state, n, m, end_point, 10000, h1)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

*****weighted AStar-h1-alpha=10000*****
execution time:  4.460335 ms
number_of_states:  633
number_of_unique_states:  231
path cost:  34
path:  RRRRRRRLDLDDDLLDLLLRDDURURURRDRRDD


|            |Path cost|Number of states|Number of unique states|Execution time|
|:-----------|:-------:|:--------------:|:---------------------:|:------------:|
|BFS         |34       |992             |350                    |8.36          |
|IDS         |34       |36067           |359                    |216.36        |
|A\*         |34       |986             |346                    |8.45          |
|weighted A\*|34       |633             |231                    |7.37          |

Test case 03 :

In [88]:
if __name__ == '__main__':
    n, m, start_point, end_point, orcs, fellowships_of_the_ring_locations = get_input()
    table = init_table(n, m, orcs)

    state = start_point, fellowships_of_the_ring_locations, ()

10 10
0 0
9 9
6 4
2 4 2
2 8 1
4 0 2
6 3 1
7 1 1
7 7 3
0 7
9 1
4 5
3 2
9 1
0 9
4 9
8 0


In [ ]:
print("******************BFS******************")
tic = time.time()
node, number_of_unique_states, number_of_states = BFS(table, state, n, m, end_point)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

In [ ]:
print("******************IDS******************")
tic = time.time()
node, number_of_unique_states, number_of_states = IDS(table, state, n, m, end_point)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

In [ ]:
print("****************AStar-h1***************")
tic = time.time()
node, number_of_unique_states, number_of_states = AStar(table, state, n, m, end_point, 1, h1)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

In [ ]:
print("*****weighted AStar-h1-alpha=10000*****")
tic = time.time()
node, number_of_unique_states, number_of_states = AStar(table, state, n, m, end_point, 10000, h1)
toc = time.time()
if not isinstance(node, Node) :
    print('Path not found!!')
else:
    print_output(tic, toc, node, number_of_unique_states, number_of_states)

با توجه آرايش اورك‌ها تست كيس 03 جوابي ندارد و برنامه پيام Path not found را در خروجي چاپ خواهد كرد. 